pkg/registry/common/memory/nse_server_test.go

```diff
@@ -177,6 +183,180 @@ func TestNetworkServiceEndpointRegistryServer_RegisterAndFindByLabelWatch(t *tes
 	require.True(t, proto.Equal(expected, <-ch))
 }
 
+func TestNetworkServiceEndpointRegistryServer_DataRace(t *testing.T) {
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s := memory.NewNetworkServiceEndpointRegistryServer()
+
+	go func() {
+		for ctx.Err() == nil {
+			nse, err := s.Register(ctx, &registry.NetworkServiceEndpoint{Name: "nse-1"})
+			require.NoError(t, err)
+
+			_, err = s.Unregister(ctx, nse)
+			require.NoError(t, err)
+		}
+	}()
+
+	c := adapters.NetworkServiceEndpointServerToClient(s)
+
+	wg := new(sync.WaitGroup)
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			stream, err := c.Find(ctx, &registry.NetworkServiceEndpointQuery{
+				NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{Name: "nse-1"},
+				Watch:                  true,
+			})
+			assert.NoError(t, err)
+
+			for j := 0; j < 100; j++ {
+				nse, err := stream.Recv()
+				assert.NoError(t, err)
+
+				nse.Name = ""
+			}
+		}()
+	}
+
+	wg.Wait()
+}
+
+func TestNetworkServiceEndpointRegistryServer_SlowReceiver(t *testing.T) {
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s := memory.NewNetworkServiceEndpointRegistryServer()
+
+	c := adapters.NetworkServiceEndpointServerToClient(s)
+
+	findCtx, findCancel := context.WithCancel(ctx)
+
+	stream, err := c.Find(findCtx, &registry.NetworkServiceEndpointQuery{
+		NetworkServiceEndpoint: &registry.NetworkServiceEndpoint{Name: "nse-1"},
+		Watch:                  true,
+	})
+	require.NoError(t, err)
+
+	for i := 0; i < 100; i++ {
+		var nse *registry.NetworkServiceEndpoint
+		nse, err = s.Register(ctx, &registry.NetworkServiceEndpoint{Name: "nse-1"})
+		require.NoError(t, err)
+
+		_, err = s.Unregister(ctx, nse)
+		require.NoError(t, err)
+	}
+
+	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
+
+	_, err = stream.Recv()
+	require.NoError(t, err)
+
+	findCancel()
+}
+
+func TestNetworkServiceEndpointRegistryServer_ShouldReceiveAllRegisters(t *testing.T) {
```

Comment:
Also works fine on master with increasing buffer size inside the test...

Reply:
Yes, setting unlimited buffer size fixes almost all found issues :)